scipy/integrate/_tanhsinh.py

```diff
@@ -760,3 +760,381 @@ def _tanhsinh_iv(f, a, b, log, maxfun, maxlevel, minlevel,
+    simplicity of exposition, the sum of a monotone decreasing function is bounded by
+
+    .. math::
+
+        \int_u^\infty f(x) dx \leq \sum_{k=u}^\infty f(k) \leq \int_u^\infty f(x) dx + f(u)
+
+    Let :math:`a` represent  `a`, :math:`n` represent `maxterms`, :math:`\epsilon_a`
+    represent `atol`, and :math:`\epsilon_r` represent `rtol`.
+    The implementation first evaluates the integral :math:`S_l=\int_a^\infty f(x) dx`
+    as a lower bound of the infinite sum. Then, it seeks a value :math:`c > a` such
+    that :math:`f(c) < \epsilon_a + S_l \epsilon_r`, if it exists; otherwise,
+    let :math:`c = a + n`. Then the infinite sum is approximated as
+    
+    .. math::
+
+        \sum_{k=a}^{c-1} f(k) + \int_c^\infty f(x) dx + f(c)/2
+
+    and the reported error is :math:`f(c)/2` plus the error estimate of
+    numerical integration. The approach described above is generalized for non-unit
+    `step` and finite `b` that is too large for direct evaluation of the sum,
+    i.e. ``b - a + 1 > maxterms``.
+
+    References
+    ----------
+    [1] Wikipedia. "Integral test for convergence."
+    https://en.wikipedia.org/wiki/Integral_test_for_convergence
+
+    Examples
+    --------
+    Compute the infinite sum of the reciprocals of squared integers.
+    
+    >>> import numpy as np
+    >>> from scipy.integrate._tanhsinh import _nsum
+    >>> res = _nsum(lambda k: 1/k**2, 1, np.inf, maxterms=1e3)
+    >>> ref = np.pi**2/6  # true value
+    >>> res.error  # estimated error
+    4.990014980029223e-07
+    >>> (res.sum - ref)/ref  # true error
+    -1.0101760641302586e-10
+    >>> res.nfev  # number of points at which callable was evaluated
+    1142
+    
+    Compute the infinite sums of the reciprocals of integers raised to powers ``p``.
+    
+    >>> from scipy import special
+    >>> p = np.arange(2, 10)
+    >>> res = _nsum(lambda k, p: 1/k**p, 1, np.inf, maxterms=1e3, args=(p,))
+    >>> ref = special.zeta(p, 1)
+    >>> np.allclose(res.sum, ref)
+    True
+    
+    """ # noqa: E501
+    # TODO
+    # - discuss how to make this do what the user intends when b is slightly
+    #   less than a plus an integer multiple of step. This needs a careful
+    #   review of the places this could get us into trouble.
+
+    # Potential future work:
+    # - add other methods for convergence acceleration (Richardson, epsilon)
+    # - support infinite lower limit?
+    # - support negative monotone increasing functions?
+    # - b < a or negative step?
+    # - complex-valued function?
+    # - look out for violations of monotonicity
+
+    # Function-specific input validation / standardization
+    tmp = _nsum_iv(f, a, b, step, args, log, maxterms, atol, rtol)
+    f, a, b, step, valid_abstep, args, log, maxterms, atol, rtol = tmp
+
+    # Additional elementwise algorithm input validation / standardization
+    tmp = _scalar_optimization_initialize(f, (a,), args, complex_ok=False)
```

Comment:
Yeah. I planned to do that. I will move all the elementwise algorithm machinery out of `_zeros.py` if you'd be willing to review that PR. (Should just be easy, of course, since the goal is to just move it without changing any behavior.)